pageserver/src/http/openapi_spec.yml

```diff
@@ -93,9 +93,16 @@ paths:
             application/json:
               schema:
                 $ref: "#/components/schemas/Error"
+        "503":
+          description: Temporarily unavailable, please retry.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ServiceUnavailableError"
```

Comment:
I've slept with it and I don't think this is right. Basically 503, 502 and some other errors could be returned by any network intermediary (LB, proxy, ingress, etc.). So specifying some specific response in the spec is just misleading generally, as there is close to zero guarantee that it response will obey this structure since it could be an HTML from AWS if they have some network problem

With pageserver it's less of the problem as we connect to EC2 directly using a DNS name, but it's true for almost any other Neon service and may change with pageservers too in the future

That said, I don't think that picking a 5xx errors for the purpose of service-related business logic errors is a good idea

I'd pick something in the 4xx range like `412 Precondition Failed` + return additional code in json if you want to distinguish different reasons of why request cannot be processed at this moment

@jcsp @koivunej

Reply:
For cases like calling into the service while it's still starting up, I think 503 will be the long term answer: the service really is unavailable.  For some other cases where an issue is local to a tenant we might evolve this to use a more expressive 409 instead -- but the move from 500 to 503 is still a net improvement as it enables us to distinguish real internal server errors from the current stream of not-really-an-error 500 cases.

I agree that clients shouldn't expect a particular body from these status codes: I've just included one here to reflect what we actually return, but my recommendation for the client would be to not count on it being there.

> Basically 503, 502 and some other errors could be returned by any network intermediary (LB, proxy, ingress, etc.)

Right, indeed -- IMO this is the main motivation to ensure the client does not rely on swagger for these, and instead have its own generic handling: https://github.com/neondatabase/cloud/issues/7238 -- that way the swagger file doesn't have to know if there's e.g. a gateway in between that might insert a 502.